Towards first complete implementation

ruairimoran opened the pull request:
## TODOs

- [x] Cholesky factoriser and solver
- [x] Transpose of tensor (i.e., transpose of each matrix)
- [x] SVD for tensors (using a loop)
- [x] Nullspace matrix and 
- [x] API docs
- [x] Update `README.md`

**Note:** In this implementation I assumed that the user provides the data in column-major format. I guess we can keep the transformation from row-major to column-major outside `DTensor`.



---

*Copied in from #15*

## Main idea

Instead of vectors, matrices, and tensors, everything is a tensor and all operations are defined on 3D tensors, $T=(T_1, T_2, \ldots, T_k)$, where $T_i\in\mathbb{R}^{m\times n}$. For exampe, $cT$, $T+T'$ and $T-T'$ are defined naturally. We also have the tensor-tensor product $TT' = (T_1T_1', \ldots, T_kT_k')$, which is performed by `cublas<t>gemmBatched`.

I have also implemented the least-squares method for tensors that allows us to do $T^\dagger B = (T_1^\dagger B_1, \ldots, T_k^\dagger B_k)$ (many least squares in parallel).

I implemented a "slicing" constructor for tensors that allows us to select matrices, columns and rows (T&C's apply of course).


### A few examples

I haven't written any proper API docs or human-readable documentation yet, but here are a few quick examples:

<details><summary>Slicing</summary>

Here we select matrices 0 and 1, so we obtain a (2,3,2)-ensor:

```c++
#define TENSOR_DATA_234A {1, 2,  3, 4, 5, 6, 7, 8, 9, 8, 7, 10, 5, 4, 3, 2, 1, -1, 4, 3, 4, 3, 4, 8}

std::vector<T> data = TENSOR_DATA_234A;
DTensor<T> tens(data, 2, 3, 4); // tensor 2 x 3 x 4
DTensor<T> tensSlice(tens, 2, 0, 1); // matrices #0 and #1
```

Here we select the columns 1 and 2 only, and we obtain a (2,2,1)-tensor

```c++
DTensor<T> tenzSlice(tens, 1, 1, 2); // columns from 1 to 2
```

We can also slice to obtain a vector of elements as follows; here the result is a (2, 1, 1)-tensor:

```c++
DTensor<T> tensSliceVec(tenz, 0, 2, 3); // elements 2..3
```
</details>

<details><summary>Uploading, downloading, d2d copying</summary>
Like this:

```c++
DTensor<T> tens(2, 3, 4);
tens.upload(data);
```

and 

```c++
DTensor<T> tens(data, 2, 3, 4);
DTensor<T> other(2, 3, 5, true); // true: zero tensor
DTensor<T> z(other, 2, 1, 4); // slicing
tens.deviceCopyTo(z);
```
</details>

<details><summary>Norms</summary>

The Frobenius norm of a tensor (the Euclidean norm when it is a vector) is `.normF()`. The sum of absolute values of all elements is `.sumAbs()`.
</details>

<details><summary>Printing</summary>
Simple as:

```c++
std::cout << tens;
```
</details>

<details><summary>Accessing individual elements</summary>
This is only for debugging purposes because it moves data from the device to the host:

```c++
double element1 = tens(2, 1, 0);
double element2 = tens(2, 1); // equivalent
```
</details>

<details><summary>Some operators</summary>

The operator `*=` is implemented for scalar multiplication (e.g, `t *= 3.0`). The operators `+=`, `-=`, `+`, and `-` have been implemented. Here is an example of using `addAB`:

```c++
std::vector<T> aData = {1, 2, 3, 4, 5, 6,
                            7, 8, 9, 10, 11, 12,
                            13, 14, 15, 16, 17, 18};
std::vector<T> bData = {6, 5, 4, 3, 2, 1,
                        7, 6, 5, 4, 3, 2,
                        1, 2, 1, 5, -6, 8};
DTensor<T> A(aData, 2, 3, 3);
DTensor<T> B(bData, 3, 2, 3);
DTensor<T> C(2, 2, 3, true);
C.addAB(A, B);  // C <- AB
```

Operator `*` has also been implemented for tensor-tensor multiplication (but `addAB` might often be more a appropriate choice).
</details>

<details><summary>Least squares</summary>
Here is a quick example

```c++
std::vector<T> aData = {1, 2,
                            3, 4,
                            7, 8,
                            9, 10,
                            6, 8,
                            -9, 20};
std::vector<T> bData = {1, 1, -1, 2, 30, -80};
DTensor<T> A0(aData, 2, 2, 3);
DTensor<T> A(A0);
DTensor<T> B(bData, 2, 1, 3);
DTensor<T> sol(B);
A0.leastSquares(sol);
DTensor<T> C(2, 1, 3);
C.addAB(A, sol);
C -= B;
T nrmErr = C.normF();
```
</details>

<details><summary>Singular value decomposition</summary>
It is currently supported for *matrices* only, i.e., (m, n, 1)-tensors, although we can easily loop and do multiple SVDs. Currently, cusolver doesn't support batched SVD, but we could use [kblas](https://github.com/ecrc/kblas-gpu?tab=readme-ov-file) if necessary. Here's a little example:


```c++
std::vector<T> bData{ 1, 6, 6, 6, 6, 6, 6, 6,
                       2, 7, 7, 7, 7, 7, 7, 7,
                       3, 8, 8, 8, 8, 8, 8, 8,};
DTensor<T> B(bData, 8, 3);
Svd<T> svd(B, true, false);
EXPECT_EQ(0, svd.factorise());
auto S = svd.singularValues();
unsigned int r = svd.rank();
``` 
</details>


## Associated issues

- Closes #2
- Closes #3
- Closes #4
- Partly addresses #5
- Closes #6
- Closes #7
- Closes #8
- Closes #9
- Closes #10 
- Closes #11 
- Closes #13 
- Closes #16

Commit:
store context as pointer, pass by reference. need to update README

alphaville reviewed (approved):
Excellent. It just says we can't merge due to some conflicts, so merge `main` into `dev` first.

Commit:
changed template functions to inline to prevent multiple defintion error

Commit:
merge main

alphaville committed:
minor

alphaville committed:
messed up the branches a bit
merging, resolving conflicts

alphaville committed:
SvdEngine: typename constraint (floating point)

alphaville committed:
compute rank of matrix with SVD
kernel that counts nonzero SVs

Commit:
implement matrix-vector multiplication for float and double

Commit:
merge dev

alphaville committed:
SvdEngine computes matrix rank

alphaville committed:
rank

alphaville committed:
merge and resolve conflicts

alphaville committed:
using template<> inline

Commit:
upgrade cmake to make device.cuh a library with necessary dependencies, and create gtests file device_test

alphaville committed:
copyRows (wip)

alphaville committed:
Merge branch 'dev' of https://github.com/GPUEngineering/GPUtils into dev

Commit:
typo

Commit:
add folders: .github, ci

Commit:
fit ci to this project

alphaville committed:
operator() for vector and matrix

alphaville committed:
test SVD

alphaville committed:
Merge branch 'dev' of https://github.com/GPUEngineering/GPUtils into dev

alphaville committed:
testing indexing of vectors/matrices

alphaville committed:
getRows: implemented and tested
works for float and double
uses cublasCopy

alphaville committed:
fix memory leak in DeviceMatrix::getRows

alphaville committed:
fix memory leaks
more testing is needed

Commit:
commented out workflow - will setup runner tomorrow. added template functions for row <-> col major storage. added tests

Commit:
Merge branch 'dev' of github.com:GPUEngineering/GPUtils into dev

Commit:
refactor tests

alphaville committed:
more thorough testing revealed memory leaks and other issues
now fixed; all tests pass; no memory-related errors

alphaville committed:
Merge branch 'dev' of https://github.com/GPUEngineering/GPUtils into dev

alphaville committed:
update readme and go to bed

alphaville committed:
introduce gpuErrChk
addressing #11

alphaville committed:
more tests for DeviceMatrix

Commit:
refactor tests

Commit:
remove repeated getRows test

alphaville committed:
yeah, we'll have conflicts now

alphaville committed:
Merge branch 'dev' of https://github.com/GPUEngineering/GPUtils into dev

alphaville committed:
Merge branch 'dev' of https://github.com/GPUEngineering/GPUtils into dev

Commit:
run tests on self hosted runner

alphaville committed:
Cholesky factorisation works, but...
solve doesnt

alphaville committed:
Merge branch 'dev' of https://github.com/GPUEngineering/GPUtils into dev

Commit:
DeviceTensor created

alphaville committed:
cholesky factorisation+solution works

Commit:
added DeviceTensor pushBack for DeviceVectors

Commit:
merge dev

Commit:
added std::vector pushBack and tests

Commit:
typo

Commit:
adding << to tensor

alphaville committed:
minor

alphaville committed:
rename variable to error

alphaville committed:
minor

alphaville committed:
proposed tensor (wip)

alphaville committed:
proposed CoolTensor

alphaville committed:
proposed DeviceTensor v0

Commit:
added example folder as template project for using GPUtils

alphaville committed:
impl DeviceMatric Constructor from DeviceVector (shallow)

alphaville committed:
DeviceMatrix::+/- and tests

Commit:
add comments

Commit:
.leastSquares added to Tensor

Commit:
leastSquares test passing. needs Tensor pushBach for DeviceVector that does not SIGSEG fault with devicePointers()

Commit:
not sure how that got changed

Commit:
refactor files

Commit:
Merge branch 'dev' into feature/experimental-tensor

Commit:
refactor ls test

Commit:
stopped GPUtils targets from being built by example project cmake and reorganised cmake for GPUtils (removed unnecessary library)

Commit:
woops

Commit:
need include/CMakeLists.txt file ?

Commit:
force c++20

Commit:
will this pass?

Commit:
fixed

alphaville committed:
Merge pull request #12 from GPUEngineering/feature/experimental-tensor

Experimental device tensor

Commit:
added info to example README

Commit:
merge dev

alphaville committed:
nullspace matrix (untested)

alphaville committed:
transpose for float and double

alphaville committed:
multiply tensors

alphaville committed:
context is now a singleton

alphaville committed:
fix memory errors

Commit:
Merge branch 'dev' into f/example

Commit:
need to update tag

Commit:
Revert "need to update tag"

This reverts commit 67e57a7c309189d8555fb5c13d301b11b4682b7e.

Commit:
refactor example to suit new file names

alphaville committed:
Tenzor - getting started

About:
- Basic memory allocation for tensors
- Printing tensors works
- Upload and download

alphaville committed:
Further development of Tenzor

alphaville committed:
slicing constructor (axis)

alphaville committed:
got rid of numAllocatedElements

alphaville committed:
started unit testing

About:
- using gpuErrChk
- unit tests for all methods of tensor
- implementation of *= (scalar), +=, -=
- assignment op and copy constr
- meaningful slicing (axis)
- no memory leaks/errors

alphaville committed:
Tensor: pointerToMatrices (tested)

alphaville committed:
Tensor addAB implemented and tested
more thorough testing needed

alphaville committed:
addAB<float> implemented

alphaville committed:
first implementation of Svd

alphaville committed:
svd done and tested

alphaville committed:
rename variables

alphaville committed:
Cholesky (impl. + tested)

alphaville committed:
operator* for DTensor

alphaville committed:
getRows: impl. + testing

Commit:
example reworked for DTensor // added example executable to ci // can't stop gtest being fetched, but now prevents tests from being built in user projects (so speedier building)

ruairimoran committed:
Merge pull request #14 from GPUEngineering/f/example

Template project

alphaville committed:
move constructor for DTensor

alphaville committed:
merge and resolve conflicts
address Ruairi's comments

alphaville committed:
Merge pull request #15 from GPUEngineering/f/tensor2

Everything/everyone is a tensor, including Ruairi and myself

alphaville commented:
@ruairimoran I think you deleted `include/CMakeLists.txt`. Is it not needed?

Also, I guess we should now delete `gputils`.

alphaville committed:
Transpose, API docs, clean up code

About:
- Separate constructor definition/impl
- API docs for DTensor
- Tensor transposition: impl + test

alphaville committed:
Merge branch 'dev' of https://github.com/GPUEngineering/GPUtils into f/tensor2

ruairimoran commented:
> @ruairimoran I think you deleted `include/CMakeLists.txt`. Is it not needed?
> 
> Also, I guess we should now delete `gputils`.

it is not, and yes we should

Commit:
remove old files

alphaville committed:
SVD supports tensors

About:
- Looping to compute all SVDs
- Thoroughly test SVD (tensors)
- Remove gputils

alphaville committed:
resolve conflicts

Commit:
API docs added

Commit:
merged dev and added nullspace api docs

alphaville committed:
SVD and Nullspace

About:
- SVD: rank is tensor (tested)
- Nullspace, draft impl. (WIP, untested)

alphaville committed:
test: rank from SVD

alphaville committed:
resolve conflicts

alphaville committed:
nullspace: very rudimentary test

Commit:
added -g to nvcc flags (gives line number for memcheck error) & testing if memcheck now throws error in ci

Commit:
Merge branch 'dev' of github.com:GPUEngineering/GPUtils into dev

Commit:
ci now fails on memcheck error

alphaville committed:
Nullspace (tested)

alphaville committed:
Merge branch 'dev' of https://github.com/GPUEngineering/GPUtils into dev

alphaville committed:
complete API docs

alphaville committed:
scalar * tensor (tested)

Commit:
README updated

Commit:
Merge branch 'dev' of github.com:GPUEngineering/GPUtils into dev

Commit:
remove TODO

Commit:
typos

Commit:
typos

Commit:
typos

Commit:
typos
